File: 常见大数据项目/flink1.14-learning/src/main/java/com/aurora/datastream/OrderAnalysisTask.java
package com.aurora.datastream;

import com.aurora.source.GenerateCustomOrderSource;
import lombok.extern.slf4j.Slf4j;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.restartstrategy.RestartStrategies;
import org.apache.flink.api.common.time.Time;
import org.apache.flink.api.java.tuple.Tuple7;
import org.apache.flink.api.java.tuple.Tuple8;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.runtime.state.hashmap.HashMapStateBackend;
import org.apache.flink.runtime.state.storage.FileSystemCheckpointStorage;
import org.apache.flink.streaming.api.CheckpointingMode;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.CheckpointConfig;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.EnvironmentSettings;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.concurrent.TimeUnit;

/**
 * @author lj.michale
 * @description
 * @date 2021-08-19
 */
@Slf4j
public class OrderAnalysisTask {

    public static void main(String[] args) throws Exception {

        ParameterTool paramTool = ParameterTool.fromArgs(args);

        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
        env.enableCheckpointing(200, CheckpointingMode.EXACTLY_ONCE);
        env.getCheckpointConfig().setMinPauseBetweenCheckpoints(30000);
        env.getCheckpointConfig().setCheckpointTimeout(10000);
        env.getCheckpointConfig().setMaxConcurrentCheckpoints(1);
        env.getCheckpointConfig().enableExternalizedCheckpoints(CheckpointConfig.ExternalizedCheckpointCleanup.RETAIN_ON_CANCELLATION);
        // 缓冲区
        env.setBufferTimeout(50L);
        // 重启策略
        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(5, Time.of(30, TimeUnit.SECONDS)));
        // 状态后端-HashMapStateBackend
        env.setStateBackend(new HashMapStateBackend());
        //等价于FsStateBackend
        env.getCheckpointConfig().setCheckpointStorage(new FileSystemCheckpointStorage("file:///E:\\OpenSource\\GitHub\\bigdata-learning\\常见大数据项目\\flink1.14-learning\\checkpoint"));
        EnvironmentSettings settings = EnvironmentSettings.newInstance().useBlinkPlanner().inStreamingMode().build();
        StreamTableEnvironment tEnv = StreamTableEnvironment.create(env, settings);

        DataStreamSource<Tuple7<String, String, String, String, String, Double, String>> oredrDS = env.addSource(new GenerateCustomOrderSource());

        // 计算BuyCount指标
        SingleOutputStreamOperator<Tuple8<String, String, String, String, String, Double, Integer, String>> ds = oredrDS.map(new MapFunction<Tuple7<String, String, String, String, String, Double, String>, Tuple8<String, String, String, String, String, Double, Integer, String>>() {
            @Override
            public Tuple8<String, String, String, String, String, Double, Integer, String> map(Tuple7<String, String, String, String, String, Double, String> input) throws Exception {
                Integer buyCount = (int) (input.f5 / Double.valueOf(input.f4.toString()));
                return new Tuple8<>(input.f0, input.f1, input.f2, input.f3, input.f4, input.f5, buyCount, input.f6);
            }
        }).name("addBuyCount");

        // 添加水印






        ds.print();




        env.execute("OrderAnalysisTask");

    }


    public static long pare(String time){
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd hh:mm:ss");
        long s = 0;
        try {
            s = sdf.parse(time).getTime();
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return s;
    }

}

File: 常见大数据项目/flink1.13-learning/src/main/scala/com/bigdata/task/example/OrderAnalysisTask.scala
package com.bigdata.task.example

import java.time.ZoneId
import java.util.Properties

import com.bigdata.bean.RawData
import com.luoj.source.GenerateCustomOrderSource
import org.apache.flink.api.common.serialization.SimpleStringSchema
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackend
import org.apache.flink.streaming.api.{CheckpointingMode, TimeCharacteristic}
import org.apache.flink.streaming.api.environment.CheckpointConfig
import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}
import org.apache.flink.streaming.connectors.kafka.{FlinkKafkaConsumer, FlinkKafkaProducer}
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer.Semantic
import org.apache.flink.table.api.EnvironmentSettings
import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment


/**
 * @author lj.michale
 * @description  Order Data Analysis
 * @date 2021-08-21
 */
object OrderAnalysisTask {

  def main(args: Array[String]): Unit = {

//    val logger:Logger = LoggerFactory.getLogger(this.getClass)

    def main(args: Array[String]): Unit = {

      val paramTool = ParameterTool.fromArgs(args)

//      logger.info(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")
      val env = StreamExecutionEnvironment.getExecutionEnvironment
      env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
      env.enableCheckpointing(200, CheckpointingMode.EXACTLY_ONCE)
      env.getCheckpointConfig.setMinPauseBetweenCheckpoints(30000)
      env.getCheckpointConfig.setCheckpointTimeout(10000)
      env.getCheckpointConfig.setMaxConcurrentCheckpoints(1)
      env.getCheckpointConfig.enableExternalizedCheckpoints(CheckpointConfig.ExternalizedCheckpointCleanup.RETAIN_ON_CANCELLATION)
      env.getCheckpointConfig.setCheckpointStorage("file:///E:\\OpenSource\\GitHub\\bigdata-learning\\常见大数据项目\\flink1.13-learning\\checkpoint")
      env.setStateBackend(new EmbeddedRocksDBStateBackend)
      val settings = EnvironmentSettings.newInstance().useBlinkPlanner().inStreamingMode().build()
      val tableEnv = StreamTableEnvironment.create(env, settings)
      tableEnv.getConfig.setLocalTimeZone(ZoneId.of("Europe/Berlin"))

      val consumerProp = new Properties()
      consumerProp.setProperty("bootstrap.servers", "localhost:9092")
      consumerProp.setProperty("group.id", "event")
      consumerProp.setProperty("key.serializer", "org.apache.kafka.common.serialization.StringSerializer")
      consumerProp.setProperty("value.serializer", "org.apache.kafka.common.serialization.StringSerializer")

      val producerProp = new Properties()
      producerProp.setProperty("bootstrap.servers", "localhost:9092")
      // 设置flink的producer的transaction.timeout.ms，flink默认为1h，大于kafka server允许的15m，会报错，所以需要修改为15m以内的时间
      producerProp.setProperty("transaction.timeout.ms", s"${1000 * 60 * 5}")

      import org.apache.flink.api.scala._
      val orderStream = env.addSource(new GenerateCustomOrderSource)
      orderStream.print()
//      // Tuple7<String, String, String, String, String, Double, String>
//      val orderDS:DataStream[String] = orderStream.map(row => {
//        val oredrId:String = row.getField(0)
//        val customerId:String = row.getField(1)
//        val productId:String = row.getField(2)
//        val productName:String = row.getField(3)
//        val price:String = row.getField(4)
//        val buyMoney:Double = row.getField(5)
//        val buyCount:Int = (buyMoney / price.toDouble).toInt
//        val buyTime:String = row.getField(5)
//        val rawData = RawData(oredrId, customerId, productId, productName, price, buyMoney, buyCount, buyTime)
//        rawData
//      }).name("getBuyCount")
//        .map(row => {
//          row.oredrId + "|" + row.customerId + "|" + row.productId + "|" + row.productName + "|" + row.price + "|" + row.buyMoney + "|" + row.buyCount + "|" + row.buyTime
//        }).name("toStringDataStream")
//
//      // 将自定义Source产生的data转发到Kafka
//      val kafkaSink = new FlinkKafkaProducer[String]("mzpns", new SimpleStringSchema(), producerProp)
//      orderDS.addSink(kafkaSink)
//
//      // consumer data from kafka
//      val kafkaSource = new FlinkKafkaConsumer[String]("mzpns", new SimpleStringSchema(), consumerProp)
//      val kafkaDataStream:DataStream[String] = env.addSource(kafkaSource).setParallelism(1)
//      kafkaDataStream.print()


      env.execute(s"${this.getClass.getName}")

    }

  }

}
